add get_data_type(), a data type detection function for severity pipeline

Attalowe opened the pull request:
## Purpose

This PR adds `get_data_type()`, a utility function to automatically classify input data for severity analyses. It supports:
- Count data (via `total_count`/`total_death` arguments)
- `incidence`-class objects
- Linelist/incidence data frames
- Error handling for unrecognized formats

## Changes
- New function in `R/get_data_type.R`
- Case-insensitive column name checks
- Validation logic for required columns (`date`, `cases`, `dead`)
- Linelist keyword matching (e.g., "id", "onset", "outcome")
- Roxygen2 documentation with other files (to be push later)

## Testing
- Includes validation for:
  - Count data detection
  - `incidence` object handling
  - Data frame type classification (linelist vs. incidence)
  - Error cases

## Notes
- usethis and devtools was use.

Attalowe committed:
feat: add data type detection function for severity pipeline

- Add `get_data_type()` to automatically classify input data as "count_data",
  "incidence", or "linelist" based on
 structure/columns
- Checks for count data via `total_count`/`total_death` arguments
- Identifies `incidence` class objects
- Validates data frames against required columns (`date`, `cases`, `dead`)
  and linelist keywords (e.g., "id", "onset", "outcome")
- Implements case-insensitive column name matching
- Includes error handling for unrecognized formats

Attalowe committed:
- Exported for package use with roxygen2 documentation get_data_type

Attalowe committed:
intial

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+
+
+#' Determine Input Data Type for Severity Pipelinele
```
```suggestion
#' Determine input data type for the severity pipeline
```

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+
+#' Determine Input Data Type for Severity Pipelinele
+#'
+#' @param data A dataframe, incidence object, or NULL.
```
```suggestion
#' @param data A dataframe-like object that can be either data.frame, linelist or incidence. Default is NULL.
```

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+#' Determine Input Data Type for Severity Pipelinele
+#'
+#' @param data A dataframe, incidence object, or NULL.
+#' @param total_count Numeric total cases (optional).
```
```suggestion
#' @param total_count A numeric with the total number of cases. Default is NULL.
```

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+#'
+#' @param data A dataframe, incidence object, or NULL.
+#' @param total_count Numeric total cases (optional).
+#' @param total_death Numeric total deaths (optional).
```
```suggestion
#' @param total_death A numeric with the total number of deaths. Default is NULL.
```

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+#' @param total_count Numeric total cases (optional).
+#' @param total_death Numeric total deaths (optional).
+#'
+#' @returns One of "count_data", "linelist", "incidence".
```
```suggestion
#' @returns A character with one of the following values: "count_data", "linelist" or "incidence".
```

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+#' @returns One of "count_data", "linelist", "incidence".
+#' @export
+#'
+#' @examples
```
Please add an example

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+#'
+#' @examples
+get_data_type <- function( data = NULL, total_count = NULL, total_death = NULL ){
+
```
use {checkmate} to check the values provided in each argument.

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+get_data_type <- function( data = NULL, total_count = NULL, total_death = NULL ){
+
+  # Check for count data
+  if ( !is.null(total_count) && !is.null(total_death)){
```
```suggestion
count_data <- c(total_count, total_death)
  if (length(count_data) == 2){
```

Attalowe replied:
refactor: Remove redundant length check for `count_data`

- The check `length(count_data) == 2` is redundant because `total_count` and `total_death` 
are already validated to be non-NULL in the initial condition.
- Checkmate ensures `total_count` and `total_death` are scalars, so `count_data` will always have length 2 if the initial condition passes.

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  # Default/error case
+  stop("unknown_data_type Provide count args, linelist, or incidence.")
```
```suggestion
  stop("Unknown data type! Either provide a value for total_count and total_death arguements or  only a value for data arguement.")
```

Karim-Mane reviewed (changes requested):
great work!

Please add the tests files as well.

Attalowe committed:
Update documentation for clarity

Attalowe committed:
use checkmate to check if 'data' is a data frame or NULL

Attalowe committed:
checkmate used to Check if 'total_count' is a single numeric value or NULL

Attalowe committed:
added checkmate to  Check if 'total_death' is a single numeric value or NULL

Attalowe committed:
checkmate use to ensure 'total_count' and 'total_death' are non-negative if provided as arg

Attalowe committed:
Added validation to ensure `total_count` and `total_death` are **both provided** (non-NULL)
  before classifying data as "count_data"

Attalowe committed:
Remove redundant length check for count_data

Attalowe committed:
Added an example and rewrote checkmate functions for clarity.

Attalowe committed:
fix: simplify input validation for `get_data_type()` using `checkmate::check_class`

- Replaced complex `checkmate::assert()` logic with a single `check_class()` call.
- Ensures `data` input can be `data.frame`, `linelist`, or `incidence` objects (or `NULL`).
- Maintains consistent validation behavior while reducing code complexity.

Attalowe committed:
Input validation with checkmate updated to recognize 'linelist' and 'incidence' classes

Attalowe committed:
count_data test options added

Attalowe committed:
updated

Attalowe committed:
incidence test options added

Attalowe committed:
linelist  test options added

Attalowe committed:
added: error message test and name edge cases

Attalowe committed:
linelist test fix

Attalowe committed:
deleted: fail test

Attalowe commented:
@Karim-Mane I've addressed the requested changes. Could you please take another look?  thank you.

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,119 @@
+#' get_data_type(total_count = 1500, total_death = 75)  # Returns "count_data"
+#'
+#' # 2. INCIDENCE OBJECT EXAMPLE
+#' # When using the incidence package's format
```
```suggestion
#' # When using an incidence object
```

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,119 @@
+  )
+
+  # Check if 'total_count'  and total_death' are a single numeric value or NULL
+  checkmate::assert_number(total_count, null.ok = TRUE)
```
add an extra argument here to specify that `total_count` and `total_death` should not be negative.

Karim-Mane commented on R/get_data_type.R:
```diff
@@ -0,0 +1,119 @@
+
+
+# check  'total_count' and 'total_death' are non-negative
+  if (!is.null(total_count)) checkmate::assert_number(total_count, lower = 0)
```
This is not necessary. Just add `lower = 0` in the above check.

Karim-Mane commented on tests/testthat/test-get_data_type.R:
```diff
@@ -0,0 +1,99 @@
+    date = Sys.Date(),
+    cases = 10,
+    dead = 2,
+    patient_age = 35,
```
```suggestion
    age = 35,
```

Karim-Mane reviewed (approved):
Well done - I left few comments on it.

It looks great to me and ready for merge after fixing these minor issues.

Attalowe committed:
Update comment to clarify incidence object usage

Attalowe committed:
Simplify number assertions by combining null and lower bound checks

Attalowe committed:
Update test variable name from patient_age to age

Attalowe commented:
@Karim-Mane All issues are addressed, and ready for your final review! 🚀

Karim-Mane reviewed (approved):
Looks good to me

Karim-Mane commented:
Please use `Rebase and merge` instead of `Squash and merge`

Attalowe commented:
@Karim-Mane  done
